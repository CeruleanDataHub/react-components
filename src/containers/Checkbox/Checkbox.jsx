import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { Icon } from "../Icon/Icon";

const HiddenCheckbox = styled.input.attrs({ type: "checkbox" })`
  border: 0;
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;

const StyledCheckbox = styled.div`
  display: inline-block;
  width: 1em;
  height: 1em;
  border: 1px solid ${({ checked, theme }) => (checked ? theme.green : "#999")};
  background: ${({ checked, theme }) =>
    checked ? theme.green : "transparent"};
  text-align: center;
  border-radius: 4px;
  line-height: 1em;
`;

const CheckmarkIcon = styled.span`
  color: ${({ theme }) => theme.checkmark};
  font-size: 0.75em;
  font-weight: bold;
`;

const CheckboxLabel = styled.span`
  color: #999;
  margin-left: 1.5em;
`;

const CheckboxContainer = styled.label`
  display: inline-flex;
  cursor: pointer;
  margin: 0.25em 0;

  &:hover {
    ${StyledCheckbox} {
      border-color: #333;
    }

    ${CheckboxLabel} {
      color: #666;
    }
  }
`;

export const Checkbox = ({ checked, value, label, onChange }) => (
  <CheckboxContainer>
    <HiddenCheckbox
      checked={checked}
      value={value}
      onChange={() => onChange(value)}
    />
    <StyledCheckbox checked={checked}>
      {checked && <Icon name="checkmark" as={CheckmarkIcon} />}
    </StyledCheckbox>
    <CheckboxLabel>{label}</CheckboxLabel>
  </CheckboxContainer>
);

Checkbox.propTypes = {
  /** Defines if the checkbox is checked or not */
  checked: PropTypes.bool,
  /** Label text */
  label: PropTypes.string,
  /** Value of the checkbox item */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Handler function */
  onChange: PropTypes.func
};

Checkbox.defaultProps = {
  checked: false,
  label: "",
  value: undefined,
  onChange: () => null
};
