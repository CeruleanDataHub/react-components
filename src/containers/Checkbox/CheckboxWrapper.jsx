import PropTypes from "prop-types";
import React, { useState } from "react";

import { Checkbox } from "./Checkbox";

export const CheckboxWrapper = ({ onToggle, label }) => {
  const [checked, setChecked] = useState(false);

  const onCheckboxChange = event => {
    setChecked(event.target.checked);
    onToggle(event.target.checked);
  };

  return (
    <Checkbox checked={checked} label={label} onChange={onCheckboxChange} />
  );
};

CheckboxWrapper.propTypes = {
  onToggle: PropTypes.func,
  label: PropTypes.string
};

CheckboxWrapper.defaultProps = {
  onToggle: () => {},
  label: ""
};
