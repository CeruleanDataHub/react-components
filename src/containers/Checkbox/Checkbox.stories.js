import { action } from "@storybook/addon-actions";
import React from "react";

import { Checkbox } from "./Checkbox";
import { CheckboxWrapper } from "./CheckboxWrapper";

export default {
  title: "Checkbox",
  component: Checkbox,
};

const checkboxes = [...Array(10).keys()].map((index) => index + 1);

export const CheckboxDefault = () => (
  <Checkbox checked={false} label="Checkbox not checked" onChange={() => {}} />
);

export const CheckboxChecked = () => (
  <Checkbox checked label="Checkbox checked" onChange={() => {}} />
);

export const CheckboxToggle = () => (
  <CheckboxWrapper
    onToggle={action("checkbox-change")}
    label="Checkbox toggle"
  />
);

export const MultipleCheckBoxes = () => (
  <div style={{ display: "flex", flexDirection: "column" }}>
    {checkboxes.map((index) => (
      <CheckboxWrapper
        onToggle={action(`checkbox-${index}-change`)}
        label={`Checkbox ${index}`}
      />
    ))}
  </div>
);
