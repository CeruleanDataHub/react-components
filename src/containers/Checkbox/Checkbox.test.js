import { mount } from "enzyme";
import React from "react";
import { Checkbox } from "./Checkbox";

describe("Checkbox", () => {
  let component;
  let onChangeMock;

  beforeEach(() => {
    onChangeMock = jest.fn();

    component = mount(
      <Checkbox value={1} label="Some label" onChange={onChangeMock} />
    );
  });

  it("renders", () => {
    expect(component).toMatchHtmlSnapshot();
  });

  it("should not call onChange yet", () => {
    expect(onChangeMock).not.toHaveBeenCalled();
  });

  it("onChange, when called, callback is called", () => {
    component
      .find("input[data-checkbox-test]")
      .props()
      .onChange();

    expect(onChangeMock).toHaveBeenCalled();
  });

  it("given checkbox is checked, is checked", () => {
    component = mount(<Checkbox onChange={onChangeMock} checked />);

    const actual = component.find("input[data-checkbox-test]");

    expect(actual).toBeChecked();
  });
});
