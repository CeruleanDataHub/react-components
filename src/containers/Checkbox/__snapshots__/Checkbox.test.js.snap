// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render Checkbox 1`] = `
<styled.label
  className=""
>
  <styled.input
    checked={false}
    onChange={[Function]}
  />
  <styled.div
    checked={false}
  />
  <styled.span />
</styled.label>
`;

exports[`Checkbox should render Checkbox with checked prop 1`] = `
<styled.label
  className=""
>
  <styled.input
    checked={true}
    onChange={[Function]}
  />
  <styled.div
    checked={true}
  >
    <Icon
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "baseHash": 1845756808,
            "componentId": "sc-AxhCb",
            "isStatic": false,
            "rules": Array [
              "
  color: black;
  font-size: 0.75em;
  font-weight: bold;
",
            ],
            "staticRulesId": "",
          },
          "displayName": "styled.span",
          "foldedComponentIds": Array [],
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "sc-AxhCb",
          "target": "span",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      name="checkmark"
    />
  </styled.div>
  <styled.span />
</styled.label>
`;
